receive_stream.go

```diff
@@ -163,13 +163,10 @@ func (s *receiveStream) readImpl(p []byte) (bool /*stream completed */, int, err
 			return false, bytesRead, fmt.Errorf("BUG: readPosInFrame (%d) > frame.DataLen (%d) in stream.Read", s.readPosInFrame, len(s.currentFrame))
 		}
 
-		s.mutex.Unlock()
-
```

Comment:
As I have described in https://github.com/lucas-clemente/quic-go/issues/2717#issuecomment-696692143, there's no need to protect `currentFrame` by a mutex.

The unlocking before making the copy here is deliberate, as copying O(1000) bytes might take a while, during which we don't want to block other Go routines that might be waiting on the mutex.

Reply:
Ah, so the mutex is protecting something else?

Reply:
Yes, it's protecting the 4 bools here:
https://github.com/lucas-clemente/quic-go/blob/27448c1fa520eddeda78cde02566d8b219963ba7/receive_stream.go#L43-L46

For example, the stream can get reset while the application is reading from it (either when `CancelRead` is called, or a RESET_STREAM frame is received). Similar for closing of the stream.

Reply:
Anyways, I dropped that particular change from this change.

Reply:
Note, I'm not completely convinced about the perf implications. A copy of 1000 bytes could end up being faster than a mutex unlock/lock; but I haven't looked done any detailed analysis... however, if it's needed for fairness, then it does make sense.

Reply:
My argument was based on performance considerations alone, not on fairness.
If copying is indeed faster than unlocking and locking the mutex, then this would be a fine change.

Reply:
If you want to write a micro-benchmark for this, we'd really appreciate that!
Please open a new issue / PR for this, and let's keep this PR for the changes to the `client`.

Reply:
What's the most likely call to be also waiting on the mutex? I'm still getting familiar with the codebase. I only noticed CloseRemote/SetReadDeadline/CancelRead, but those don't seem like things that are called frequently.

Reply:
You're right about that. The other calls are mostly just called once in the lifetime of a stream (maybe a few times if we receive a retransmission for a RESET_STREAM frame). The only thing that's called more often would be `SetReadDeadline`, but that should also be relatively infrequent.

Reply:
In that case, there wouldn't be contention on the mutex and probably there wouldn't be a significant difference in perf - i.e. it can use mutex fast-path. But, yeah, the only reason I can see to keep the mutex is to get the information from those methods earlier into the loop -- e.g. if there are bunch of frames in the queue, then I think it would need to exhaust them or fill `p`.

Either way, I'm convinced there isn't a significant problem with it.